File: android/src/main/java/expo/modules/deviceinfo/ExpoDeviceInfoModule.kt
package expo.modules.deviceinfo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL

class ExpoDeviceInfoModule : Module() {
  override fun definition() = ModuleDefinition {

    Name("ExpoDeviceInfo")

    Constants(
      "PI" to Math.PI
    )

    Events("onChange")

    Function("getOperatingSystemVersion") {
      return@Function android.os.Build.VERSION.RELEASE
    }

    Function("getModel") {
      return@Function android.os.Build.MODEL
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    View(ExpoDeviceInfoView::class) {
      Events(arrayOf("onLoadingFinished", "onLoadingStarted"))

      Prop("url") { view: ExpoDeviceInfoView, url: URL? ->
        view.webView.loadUrl(url.toString())
      }
    }
  }
}

File: ios/ExpoDeviceInfoModule.swift
import ExpoModulesCore

public class ExpoDeviceInfoModule: Module {
  public func definition() -> ModuleDefinition {
      
    Name("ExpoDeviceInfo")

    Constants([
      "PI": Double.pi
    ])

    Events("onChange")

    Function("getOperatingSystemVersion") { () -> String in
        let systemVersion = UIDevice.current.systemVersion
        return systemVersion
    }
      
    Function("getModel") { () -> String in
        let deviceModel = UIDevice.DeviceMap.modelName
        return deviceModel
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }
      
    View(ExpoDeviceInfoView.self) {
        Events(["onLoadingFinished", "onLoadingStarted"])
        
        Prop("url") { (view, url: URL) in
                if view.webView.url != url {
                  let urlRequest = URLRequest(url: url)
                  view.webView.load(urlRequest)
                }
              }
    }
  }
}
